### ISSUE TITLE
Ability to get/know the type of object for which the factory was called

### ISSUE THREAD

**Issue by Neakita** [DevTeam/Pure.DI#76]
Greetings!

I have a situation where I want to get an ILogger (from Serilog) dependency in a certain class:

```
internal sealed class WriteableBitmapPool : IDisposable
{
	...
	public WriteableBitmapPool(ILogger logger)
	{
		_logger = logger;
	}
	...
}
```

Ideally, the default factory should provide a contextual logger, i.e. it should be obtained from the `Log.ForContext<T>()` or `Log.ForContext(typeof(T))` method, so when `WriteableBitmapPool` is being created it should get logger instance from result of calling `Log.ForContext<WriteableBitmapPool>()` method.

Perhaps the API should look something like
```
private void Setup() => DI.Setup(nameof(Composition))
		...
		.Bind<ILogger>().To(context => Log.ForContext<TSomething>())
		...
```
_where `TSomething` is replaced with something more suitable_

so that the generator can convert this to something like
`Serilog.ILogger logger = Log.ForContext<WriteableBitmapPool>();`
just before `WriteableBitmapPool` instance creation

Alternatively add some `Type` property to `IContext` so factory could look like
`.Bind<ILogger>().To(context => Log.ForContext(context.ObjType))`

**Comment by NikolayPianikov** [comment]
The problem is that one object can be used as a dependency to create objects of different types. This depends, for example, on the lifetime. I made [an example](https://github.com/DevTeam/Pure.DI/blob/master/readme/serilog.md) of how I would implement logging.

**Comment by Neakita** [comment]
> The problem is that one object can be used as a dependency to create objects of different types. This depends, for example, on the lifetime.

I agree that this is a problem with an ambiguous solution, but in my opinion it can be solved by having each specific `ILogger<T>` have its own lifetime, i.e. if we register a singleton `ILogger<T>` that can be obtained by requesting `ILogger` in the constructor, all objects of the `Dependency` type will receive their single instance of `Logger<Dependency>`, and all objects of the `Service` type will receive their single instance of `Logger<Service>`. I believe this is how generic `Dependency<TT>` work.

> I made [an example](https://github.com/DevTeam/Pure.DI/blob/master/readme/serilog.md) of how I would implement logging.

The solution with creating additional types representing loggers for specific types certainly has a place to be, but it creates unnecessary verbosity and problems in cases where, for example, it is necessary to pass a non-default logger.

Also, apparently, in this example there is a composition configuration [error](https://github.com/DevTeam/Pure.DI/blob/master/readme/serilog.md?plain=1#L50) (it breaks off in mid-sentence)

**Comment by NikolayPianikov** [comment]
@Neakita could you please review [this approach](https://github.com/DevTeam/Pure.DI/blob/master/readme/serilog.md)?

**Comment by Neakita** [comment]
@NikolayPianikov

This is perfect for my case, however, the choice of name and the fact that the meaning depends on lifetime can be confusing, although I can't think of something more appropriate.

Unless, judging by the name, the dependency is given to the object in ownership, which I would consider a bad approach, because then it should take responsibility for managing the lifetime of its dependencies, which it can only do if it creates them itself, although in this case it gets them from outside and the owner is considered a Composition that should manage their lifetime, there is an inconsistency here when the lifetime is, for example, transient.

**Comment by NikolayPianikov** [comment]
@Neakita Yes that's a reasonable point. Please see [this option](https://github.com/DevTeam/Pure.DI/blob/master/readme/serilog.md).

**Comment by Neakita** [comment]
@NikolayPianikov 

As far as I understand, in this example you replaced `OwnerType` with `ConsumerTypes`, where in the case of Singleton or PerResolve dependency there will be several types that use it.

It seems to me that the [previous](https://github.com/DevTeam/Pure.DI/blob/ea62e5a51a8a4a9bfd38b48bfe1cae0afc0083d8/readme/serilog.md) example providing the `OwnerType` property was more understandable and easier to use, whereas the option with `ConsumerTypes` looks more flexible, I can't think of ways to use it at the moment. However, it seems to me that the `Consumer` looks more appropriate naming than the `Owner`.

My point was that, as you said in your [report](https://youtu.be/mr9zmZhboNM?si=zIxvZcso2eq6L0qw&t=3400) on DotNext, Pure.DI is intentionally not responsible for the lifetime of dependencies of the Transient type implementing `IDisposable` because it requires a lot of overhead.

I have a suggestion about this that I would like to make separate issue later.

In my opinion, something like `ConsumerType` will work more than adequately, however, for such a scenario, open generics can also be adapted for grouping by type, so `Log.ForContext<TT>` can be a singleton for each concrete `TT`, although I suppose it is too difficult to implement for such a narrow use case, but it will also worsen performance because it will be necessary to use lock for all these resolves, additional time will be spent on compiling all generic methods, and objects will not be able to be collected by GC until Composition is disposed.

I'm sorry if I'm confusing or distracting you.

**Comment by NikolayPianikov** [comment]
> As far as I understand, in this example you replaced `OwnerType` with `ConsumerTypes`, where in the case of Singleton or PerResolve dependency there will be several types that use it.

Yes, that's the way it is

> It seems to me that the [previous](https://github.com/DevTeam/Pure.DI/blob/ea62e5a51a8a4a9bfd38b48bfe1cae0afc0083d8/readme/serilog.md) example providing the `OwnerType` property was more understandable and easier to use, whereas the option with `ConsumerTypes` looks more flexible, I can't think of ways to use it at the moment. However, it seems to me that the `Consumer` looks more appropriate naming than the `Owner`.

The array option would be useful when we want to account for all consumers.

> My point was that, as you said in your [report](https://youtu.be/mr9zmZhboNM?si=zIxvZcso2eq6L0qw&t=3400) on DotNext, Pure.DI is intentionally not responsible for the lifetime of dependencies of the Transient type implementing `IDisposable` because it requires a lot of overhead.

Right now it is possible to control transient `IDisposable`. See [this](https://github.com/DevTeam/Pure.DI/blob/master/readme/tracking-disposable-instances-per-a-composition-root.md) or [this](https://github.com/DevTeam/Pure.DI/blob/master/readme/tracking-disposable-instances-in-delegates.md) example.

> In my opinion, something like `ConsumerType` will work more than adequately, however, for such a scenario, open generics can also be adapted for grouping by type, so `Log.ForContext<TT>` can be a singleton for each concrete `TT`, although I suppose it is too difficult to implement for such a narrow use case, but it will also worsen performance because it will be necessary to use lock for all these resolves, additional time will be spent on compiling all generic methods, and objects will not be able to be collected by GC until Composition is disposed.

I haven't figured out how to use context to call a method with a generic type.

**Comment by Neakita** [comment]
> I haven't figured out how to use context to call a method with a generic type.

Not a context, but [marker type such as TT, TT1, etc](https://github.com/DevTeam/Pure.DI/blob/master/readme/generics.md), some brand new `TConsumer`.
But again, if `OwnerType`/`ConsumerTypes`/`ConsumerType` will be implemented, in my case I would prefer to use it because `Log.ForContext<T>()` under the hood still calls `Log.ForContext(typeof(T))`, however, a certain `TConsumer` marker type can also be potentially useful.

**Comment by Neakita** [comment]
> The array option would be useful when we want to account for all consumers.

Not a logger case, maybe something other. I saw the generated code and it can actually be stackalloc ReadOnlySpan which prevents unnecessary array allocation.

**Comment by NikolayPianikov** [comment]
> Not a logger case, maybe something other. I saw the generated code and it can actually be `stackalloc` `ReadOnlySpan` which prevents unnecessary array allocation.

This is already implemented for some cases, e.g. see [here](https://github.com/DevTeam/Pure.DI/blob/master/readme/span-and-readonlyspan.md).

**Comment by Neakita** [comment]
> This is already implemented for some cases, e.g. see [here](https://github.com/DevTeam/Pure.DI/blob/master/readme/span-and-readonlyspan.md).

I'm talking about [this](https://github.com/DevTeam/Pure.DI/blob/master/readme/serilog.md?plain=1#L87).

**Comment by NikolayPianikov** [comment]
> Not a context, but [marker type such as TT, TT1, etc](https://github.com/DevTeam/Pure.DI/blob/master/readme/generics.md), some brand new `TConsumer`. But again, if `OwnerType`/`ConsumerTypes`/`ConsumerType` will be implemented, in my case I would prefer to use it because `Log.ForContext<T>()` under the hood still calls `Log.ForContext(typeof(T))`, however, a certain `TConsumer` marker type can also be potentially useful.

Yeah I get your idea, created a [ticket](https://github.com/DevTeam/Pure.DI/issues/77) and will think about it.

**Comment by NikolayPianikov** [comment]
Fixed in 2.1.42
